File: jambeth/jambeth-ioc/src/main/java/de/osthus/ambeth/ioc/link/ILinkController.java
package de.osthus.ambeth.ioc.link;

import de.osthus.ambeth.ioc.IServiceContext;
import de.osthus.ambeth.ioc.config.IBeanConfiguration;

/**
 * Service bean for registering a bean (extension/plugin) to extendables at runtime using a fluent interface.
 * 
 * @see ILinkExtendable
 */
public interface ILinkController
{
	/**
	 * Start linking a named bean.
	 * 
	 * @param serviceContext
	 *            IoC context containing the bean and the registry.
	 * @param listenerBeanName
	 *            Name of the bean to add to a registry.
	 * @return Next step of the fluent interface.
	 */
	ILinkRegistryNeededRuntime<?> link(IServiceContext serviceContext, String listenerBeanName);

	/**
	 * Start linking a named bean to a specific registry method.
	 * 
	 * @param serviceContext
	 *            IoC context containing the bean and the registry.
	 * @param listenerBeanName
	 *            Name of the bean to add to a registry.
	 * @param methodName
	 *            Name of the registry method.
	 * @return Next step of the fluent interface.
	 */
	ILinkRegistryNeededRuntime<?> link(IServiceContext serviceContext, String listenerBeanName, String methodName);

	/**
	 * Start linking a bean identified by its {@link IBeanConfiguration}.
	 * 
	 * @param serviceContext
	 *            IoC context containing the bean and the registry.
	 * @param listenerBean
	 *            Identifier for the bean to add to a registry.
	 * @return Next step of the fluent interface.
	 */
	ILinkRegistryNeededRuntime<?> link(IServiceContext serviceContext, IBeanConfiguration listenerBean);

	/**
	 * Start linking a bean identified by its {@link IBeanConfiguration} to a specific registry method.
	 * 
	 * @param serviceContext
	 *            IoC context containing the bean and the registry.
	 * @param listenerBean
	 *            Identifier for the bean to add to a registry.
	 * @param methodName
	 *            Name of the registry method.
	 * @return Next step of the fluent interface.
	 */
	ILinkRegistryNeededRuntime<?> link(IServiceContext serviceContext, IBeanConfiguration listenerBean, String methodName);

	/**
	 * Start linking a object or bean instance.
	 * 
	 * @param serviceContext
	 *            IoC context containing the bean and the registry.
	 * @param listener
	 *            Object or bean to add to a registry.
	 * @return Next step of the fluent interface.
	 */
	<D> ILinkRegistryNeededRuntime<D> link(IServiceContext serviceContext, D listener);

	/**
	 * Start linking a object or bean instance to a specific registry method.
	 * 
	 * @param serviceContext
	 *            IoC context containing the bean and the registry.
	 * @param listener
	 *            Object or bean to add to a registry.
	 * @param methodName
	 *            Name of the registry method.
	 * @return Next step of the fluent interface.
	 */
	ILinkRegistryNeededRuntime<?> link(IServiceContext serviceContext, Object listener, String methodName);

	/**
	 * Start linking a named bean to a specific registry method without using the fluent interface.
	 * 
	 * @param listenerBeanName
	 *            Name of the bean to add to a registry.
	 * @param methodName
	 *            Name of the registry method.
	 * @return The LinkConfiguration object.
	 */
	LinkConfiguration<Object> createLinkConfiguration(String listenerBeanName, String methodName);

	/**
	 * Start linking a bean identified by its {@link IBeanConfiguration} to a specific registry method without using the fluent interface.
	 * 
	 * @param listenerBean
	 *            Identifier for the bean to add to a registry.
	 * @param methodName
	 *            Name of the registry method.
	 * @return The LinkConfiguration object.
	 */
	LinkConfiguration<Object> createLinkConfiguration(IBeanConfiguration listenerBean, String methodName);

	/**
	 * Start linking a object or bean instance to a specific registry method without using the fluent interface.
	 * 
	 * @param listener
	 *            Object or bean to add to a registry.
	 * @param methodName
	 *            Name of the registry method.
	 * @return The LinkConfiguration object.
	 */
	LinkConfiguration<Object> createLinkConfiguration(Object listener, String methodName);

	/**
	 * Start linking a object or bean instance without using the fluent interface.
	 * 
	 * @param listener
	 *            Object or bean to add to a registry.
	 * @return The LinkConfiguration object.
	 */
	<D> LinkConfiguration<D> createLinkConfiguration(D listener);

	@Deprecated
	IBeanConfiguration createLinkConfiguration(String registryBeanName, String listenerBeanName, Class<?> registryClass);

	@Deprecated
	IBeanConfiguration createLinkConfiguration(String registryBeanName, String listenerBeanName, Class<?> registryClass, Object... arguments);

	@Deprecated
	IBeanConfiguration createLinkConfiguration(String listenerBeanName, Class<?> autowiredRegistryClass);

	@Deprecated
	IBeanConfiguration createLinkConfiguration(String listenerBeanName, Class<?> autowiredRegistryClass, Object... arguments);

	@Deprecated
	void link(IServiceContext serviceContext, String registryBeanName, String listenerBeanName, Class<?> registryClass, Object... arguments);

	@Deprecated
	void link(IServiceContext serviceContext, String registryBeanName, String listenerBeanName, Class<?> registryClass);

	@Deprecated
	void link(IServiceContext serviceContext, IBeanConfiguration listenerBean, Class<?> autowiredRegistryClass);

	@Deprecated
	void link(IServiceContext serviceContext, IBeanConfiguration listenerBean, Class<?> autowiredRegistryClass, Object... arguments);

	@Deprecated
	void link(IServiceContext serviceContext, String listenerBeanName, Class<?> autowiredRegistryClass);

	@Deprecated
	void link(IServiceContext serviceContext, String listenerBeanName, Class<?> autowiredRegistryClass, Object... arguments);
}

File: ambeth/Ambeth.IoC/ambeth/ioc/link/ILinkController.cs
using De.Osthus.Ambeth.Ioc.Config;
using System;

namespace De.Osthus.Ambeth.Ioc.Link
{

    public interface ILinkController
    {
        [Obsolete]
        IBeanConfiguration createLinkConfiguration(String registryBeanName, String listenerBeanName, Type registryClass);

        [Obsolete]
        IBeanConfiguration createLinkConfiguration(String registryBeanName, String listenerBeanName, Type registryClass, params Object[] arguments);

        [Obsolete]
        IBeanConfiguration createLinkConfiguration(String listenerBeanName, Type autowiredRegistryClass);

        [Obsolete]
        IBeanConfiguration createLinkConfiguration(String listenerBeanName, Type autowiredRegistryClass, params Object[] arguments);

        [Obsolete]
        void Link(IServiceContext serviceContext, String registryBeanName, String listenerBeanName, Type registryClass);

        [Obsolete]
        void Link(IServiceContext serviceContext, String registryBeanName, String listenerBeanName, Type registryClass, params Object[] arguments);

        [Obsolete]
        void Link(IServiceContext serviceContext, IBeanConfiguration listenerBean, Type autowiredRegistryClass);

        [Obsolete]
        void Link(IServiceContext serviceContext, IBeanConfiguration listenerBean, Type autowiredRegistryClass, params Object[] arguments);

        [Obsolete]
        void Link(IServiceContext serviceContext, String listenerBeanName, Type autowiredRegistryClass);

        [Obsolete]
        void Link(IServiceContext serviceContext, String listenerBeanName, Type autowiredRegistryClass, params Object[] arguments);

        [Obsolete]
        void Link<R>(IServiceContext serviceContext, String registryBeanName, String listenerBeanName);

        [Obsolete]
        void Link<R>(IServiceContext serviceContext, String registryBeanName, String listenerBeanName, params Object[] arguments);

        [Obsolete]
        void Link<R>(IServiceContext serviceContext, IBeanConfiguration listenerBean);

        [Obsolete]
        void Link<R>(IServiceContext serviceContext, IBeanConfiguration listenerBean, params Object[] arguments);

        [Obsolete]
        void Link<R>(IServiceContext serviceContext, String listenerBeanName);

        [Obsolete]
        void Link<R>(IServiceContext serviceContext, String listenerBeanName, params Object[] arguments);

        [Obsolete]
        IBeanConfiguration createEventLinkConfiguration(String eventProviderBeanName, Type eventInterface, String listenerBeanName, String methodName);

        [Obsolete]
        IBeanConfiguration createEventLinkConfiguration(String eventProviderBeanName, Type eventInterface, String handlerDelegateBeanName);

        [Obsolete]
        IBeanConfiguration createEventLinkConfiguration(String eventProviderBeanName, Type eventInterface, Delegate handlerDelegate);

        //[Obsolete]
        //IBeanConfiguration createEventLinkConfiguration(Type autowiredEventProviderType, String eventName, String handlerDelegateBeanName);

        [Obsolete]
        IBeanConfiguration createEventLinkConfiguration<D>(String eventProviderBeanName, IEventDelegate<D> eventName, String listenerBeanName, String methodName);

        [Obsolete]
        IBeanConfiguration createEventLinkConfiguration<D>(String eventProviderBeanName, IEventDelegate<D> eventName, String handlerDelegateBeanName);

        [Obsolete]
        IBeanConfiguration createEventLinkConfiguration<D>(String eventProviderBeanName, IEventDelegate<D> eventName, D handlerDelegate);

        //[Obsolete]
        //IBeanConfiguration createEventLinkConfiguration(Type autowiredEventProviderType, String eventName, Delegate handlerDelegate);

        [Obsolete]
        void LinkToEvent<D>(IServiceContext serviceContext, String eventProviderBeanName, IEventDelegate<D> eventName, String listenerBeanName, String methodName);

        [Obsolete]
        void LinkToEvent<D>(IServiceContext serviceContext, String eventProviderBeanName, IEventDelegate<D> eventName, String handlerDelegateBeanName);

        [Obsolete]
        void LinkToEvent<D>(IServiceContext serviceContext, String eventProviderBeanName, IEventDelegate<D> eventName, D handlerDelegate);

        [Obsolete]
        void LinkToEvent<R>(IServiceContext serviceContext, String eventProviderBeanName, String listenerBeanName, String methodName);

        [Obsolete]
        void LinkToEvent<R>(IServiceContext serviceContext, String eventProviderBeanName, String handlerDelegateBeanName);

        [Obsolete]
        void LinkToEvent<R>(IServiceContext serviceContext, String eventProviderBeanName, Delegate handlerDelegate);

        ILinkRegistryNeededRuntime Link(IServiceContext serviceContext, String listenerBeanName);

        ILinkRegistryNeededRuntime Link(IServiceContext serviceContext, String listenerBeanName, String methodName);

        ILinkRegistryNeededRuntime Link(IServiceContext serviceContext, IBeanConfiguration listenerBean);

        ILinkRegistryNeededRuntime Link(IServiceContext serviceContext, IBeanConfiguration listenerBean, String methodName);

        ILinkRegistryNeededRuntime Link(IServiceContext serviceContext, Object listener, String methodName);

        ILinkRegistryNeededRuntime<D> Link<D>(IServiceContext serviceContext, D listener);

        LinkConfiguration<Object> CreateLinkConfiguration(String listenerBeanName, String methodName);

        LinkConfiguration<Object> CreateLinkConfiguration(IBeanConfiguration listenerBean, String methodName);

        LinkConfiguration<Object> CreateLinkConfiguration(Object listener, String methodName);

        LinkConfiguration<D> CreateLinkConfiguration<D>(D listener);
    }
}
